src/puppetlabs/puppetdb/command.clj

```diff
@@ -317,18 +433,124 @@
+  "Discards the given `cmd` caused by `ex`"
+  [cmd ex q dlo]
+  (-> cmd
+      (queue/cons-attempt ex)
+      (dlo/discard-cmdref q dlo)))
+
+(defn process-delete-cmdref
+  "Processes a command ref marked for deletion. This is similar to
+  processing a non-delete cmdref except different metrics need to be
+  updated to indicate the difference in command"
+  [{:keys [command version] :as cmdref} q scf-write-db response-chan stats]
+  (process-command-and-respond! cmdref scf-write-db response-chan stats)
+  (queue/ack-command q {:entry (queue/cmdref->entry cmdref)})
+  (update-counter! :depth command version dec!)
+  (update-counter! :invalidated command version dec!))
+
+(defn process-cmdref
+  "Parses, processes and acknowledges a successful command ref and
+  updates the relevant metrics. Any exceptions that arise are
+  unhandled and expected to be caught by the caller."
+  [cmdref q scf-write-db response-chan stats]
+  (let [{:keys [command version] :as cmd} (queue/cmdref->cmd q cmdref)
+        retries (count (:attempts cmdref))]
+
+    (create-metrics-for-command! command version)
+
+    (mark-both-metrics! command version :seen)
+    (update! (global-metric :retry-counts) retries)
+    (update! (cmd-metric command version :retry-counts) retries)
+
+    (mutils/multitime!
+     [(global-metric :processing-time)
+      (cmd-metric command version :processing-time)]
+
+     (process-command-and-respond! cmd scf-write-db response-chan stats)
+     (mark-both-metrics! command version :processed))
+
+    (queue/ack-command q cmd)
+    (update-counter! :depth command version dec!)))
+
+(def command-delay-in-ms (* 1000 60 60))
+
+(defn send-delayed-message
+  "Will delay `cmd` in the `delay-pool` threadpool for
+  `command-delay-in-ms`. It will then be enqueued in `command-chan`
+  for another attempt at processing"
+  [cmd ex command-chan delay-pool]
+  (let [narrowed-entry (-> cmd
+                           (queue/cons-attempt ex)
+                           (dissoc :payload))]
+    (after command-delay-in-ms #(async/>!! command-chan narrowed-entry) delay-pool)))
+
+(defn message-handler
+  "Processes the message via (process-message msg), retrying messages
+  that fail via (delay-message msg), and discarding messages that have
+  fatal errors or have exceeded their maximum allowed attempts."
+  [q command-chan dlo delay-pool scf-write-db response-chan stats]
+  (fn [{:keys [certname command version delete? id] :as cmdref}]
+    (try+
+     (if delete?
+       (process-delete-cmdref cmdref q scf-write-db response-chan stats)
+       (let [retries (count (:attempts cmdref))]
+         (try+
+          (process-cmdref cmdref q scf-write-db response-chan stats)
+          (catch fatal? obj
+            (mark! (global-metric :fatal))
+            (let [ex (:cause obj)]
+              (log/error
+               (:wrapper &throw-context)
+               (i18n/trs "[{0}] [{1}] Fatal error on L2 attempt {2} for {3}" id command retries certname))
```

Comment:
I agree. I think the log messages look good just by removing L2, and keeping the rest, so I did that for all of them.